mindmeld/text_preparation/tokenizers.py

```diff
@@ -0,0 +1,199 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2015 Cisco Systems, Inc. and others.  All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains Tokenizers."""
+
+from abc import ABC, abstractmethod
+import logging
+import unicodedata
+
+from .spacy_model_factory import SpacyModelFactory
+
+logger = logging.getLogger(__name__)
+
+
+class Tokenizer(ABC):
+    """Abstract Tokenizer Base Class."""
+
+    def __init__(self):
+        """Creates a tokenizer instance."""
+        pass
+
+    @abstractmethod
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class NoOpTokenizer(Tokenizer):
+    """A No-Ops tokenizer."""
+
+    def __init__(self):
+        """Initialize the NoOpTokenizer."""
+        pass
+
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        return [text]
+
+
+class CharacterTokenizer(Tokenizer):
+    """A Tokenizer that splits text at the character level."""
+
+    def __init__(self):
+        """Initializes the CharacterTokenizer."""
+        pass
+
+    def tokenize(self, text):
+        """
+        Identify tokens in text and create normalized tokens that contain the text and start index.
+        Args:
+            text (str): The text to normalize
+        Returns:
+            tokens (List[str]): List of tokenized tokens.
+        """
+        tokens = []
+        token = {}
+        token_text = ""
+        for i, char in enumerate(text):
+            cat = unicodedata.category(char)
+            if char.isspace():
+                # if we hit a space, close the token and empty the buffer
+                if token and token_text:
+                    token["text"] = token_text
+                    tokens.append(token)
+                token = {}
+                token_text = ""
+                continue
+            if (
+                token_text and cat[0] != unicodedata.category(token_text[-1])[0]
```

Comment:
Please make `cat[0] != unicodedata.category(token_text[-1])[0]` more readable. What exactly is  `unicodedata.category` and why do we do this check? I see the statement about non-latin chars below, but its non-obvious how the condition causes this.

Reply:
Refactored this function and broke it down into 3 separate steps/functions to improve readability.

A new token is start if the character in sequence is one of the following:
1. It is a non-Latin Token
2. The previous token is a different unicode category. We do this because unicode categories tell us if characters are math symbols, numbers, or letters etc. We want to keep adjacent characters with the same type together. For example, numbers like "1234" in Japanese. More details at: https://www.compart.com/en/unicode/category
3. The character follows a space.

The refactor makes these steps more intuitive.